File: go-knowledge-app/internal/collectors/cost.go
package collectors

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"go-knowledge-app/internal/shared"
)

func usageDir() string {
	if env := os.Getenv("PKM_USAGE_DIR"); env != "" {
		return env
	}
	return filepath.Join(Home(), "GITHUB", "ObsidianPKM", "skripte", "usage")
}

type dailyRow struct {
	date string
	cost float64
}

func loadDailyCosts(dir string) []dailyRow {
	rows := []dailyRow{}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return rows
	}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		if strings.HasSuffix(e.Name(), "_rl.json") || strings.HasSuffix(e.Name(), ".migrated") {
			continue
		}
		text, ok := shared.ReadText(filepath.Join(dir, e.Name()))
		if !ok || text == "" {
			continue
		}
		var data map[string]any
		if json.Unmarshal([]byte(text), &data) != nil {
			continue
		}
		if _, has := data["ccusage_daily"]; has {
			continue
		}
		d := asStr(data["date"])
		if d == "" {
			continue
		}
		var cost float64
		if ccusage, ok := data["ccusage"].(map[string]any); ok {
			cost = parseFloatOrZero(asStr(ccusage["total_cost"]))
		} else {
			cost = parseFloatOrZero("0")
		}
		rows = append(rows, dailyRow{date: d, cost: cost})
	}
	return rows
}

// parseFloatOrZero mirrors `parseFloat(String(x ?? "0")) || 0`.
func parseFloatOrZero(s string) float64 {
	f, err := strconv.ParseFloat(strings.TrimSpace(s), 64)
	if err != nil || math.IsNaN(f) {
		return 0
	}
	return f
}

type rlSnapshot struct {
	Ts               string   `json:"ts"`
	FiveHourPct      *float64 `json:"five_hour_pct"`
	FiveHourResetsAt *float64 `json:"five_hour_resets_at"`
	SevenDayPct      *float64 `json:"seven_day_pct"`
	SevenDayResetsAt *float64 `json:"seven_day_resets_at"`
}

func latestRlSnapshot(dir string) rlSnapshot {
	bestTs := ""
	best := rlSnapshot{}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return best
	}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), "_rl.json") {
			continue
		}
		text, ok := shared.ReadText(filepath.Join(dir, e.Name()))
		if !ok || text == "" {
			continue
		}
		var data rlSnapshot
		if json.Unmarshal([]byte(text), &data) != nil {
			continue
		}
		if data.Ts > bestTs {
			bestTs = data.Ts
			best = data
		}
	}
	return best
}

func rateLimitFromSnapshot(pct, resetsAt *float64, windowSecs float64, nowTs int64) map[string]any {
	if pct == nil || resetsAt == nil {
		return nil
	}
	remaining := *resetsAt - float64(nowTs)
	if remaining <= 0 || remaining > windowSecs {
		return nil
	}
	elapsed := windowSecs - remaining
	if elapsed <= 0 {
		return nil
	}
	frac := elapsed / windowSecs
	projected := 0.0
	if frac > 0 {
		projected = *pct / frac
	}
	d := math.Floor(remaining / 86400)
	rem1 := math.Mod(remaining, 86400)
	h := math.Floor(rem1 / 3600)
	m := math.Floor(math.Mod(rem1, 3600) / 60)
	var countdown string
	switch {
	case d != 0:
		countdown = fmt.Sprintf("%dd%dh", int(d), int(h))
	case h != 0:
		countdown = fmt.Sprintf("%dh%dm", int(h), int(m))
	default:
		countdown = fmt.Sprintf("%dm", int(m))
	}
	return map[string]any{
		"used_pct":      math.Round(*pct*10) / 10,
		"elapsed_pct":   math.Round(frac*1000) / 10,
		"projected_pct": math.Round(projected*10) / 10,
		"resets_in":     countdown,
	}
}

// CollectCost reports account-wide $ from usage/*.json + *_rl.json. Week starts
// Friday (Claude 7d reset Thu->Fri). Pass an empty dir to use PKM_USAGE_DIR.
func CollectCost(dir string) map[string]any {
	udir := dir
	if udir == "" {
		udir = usageDir()
	}
	if _, err := os.Stat(udir); err != nil {
		return map[string]any{"available": false, "reason": fmt.Sprintf("usage/ nicht gefunden: %s", udir)}
	}
	rows := loadDailyCosts(udir)
	if len(rows) == 0 {
		return map[string]any{"available": false, "reason": fmt.Sprintf("keine ccusage-JSON in %s", udir)}
	}

	now := time.Now()
	todayStr := now.UTC().Format("2006-01-02")
	yestStr := now.Add(-24 * time.Hour).UTC().Format("2006-01-02")
	// Week starts Friday (JS Sun=0..Sat=6, Fri=5)
	dow := int(now.Weekday())
	daysFromFri := (dow - 5 + 7) % 7
	weekStr := now.Add(-time.Duration(daysFromFri) * 24 * time.Hour).UTC().Format("2006-01-02")
	monthStr := fmt.Sprintf("%04d-%02d-01", now.Year(), int(now.Month()))

	var todayC, yestC, weekC, monthC, totalC float64
	for _, r := range rows {
		totalC += r.cost
		if r.date == todayStr {
			todayC += r.cost
		}
		if r.date == yestStr {
			yestC += r.cost
		}
		if r.date >= weekStr {
			weekC += r.cost
		}
		if r.date >= monthStr {
			monthC += r.cost
		}
	}

	result := map[string]any{
		"available": true,
		"today":     todayC,
		"yesterday": yestC,
		"week":      weekC,
		"month":     monthC,
		"total":     totalC,
		"five_hour": nil,
		"seven_day": nil,
	}

	rl := latestRlSnapshot(udir)
	if rl.Ts != "" {
		nowTs := now.Unix()
		result["five_hour"] = rateLimitFromSnapshot(rl.FiveHourPct, rl.FiveHourResetsAt, 5*3600, nowTs)
		result["seven_day"] = rateLimitFromSnapshot(rl.SevenDayPct, rl.SevenDayResetsAt, 7*24*3600, nowTs)
	}

	return result
}

File: deno-knowledge-app/src/collectors/cost.ts
// collect_cost — mirrors knowledge.py collect_cost
// Account-wide $ from usage/*.json + *_rl.json (env PKM_USAGE_DIR).
// Week starts Friday (Claude 7d reset Thu->Fri).

import { readText } from "../shared.ts";
import { join } from "@std/path";

function usageDir(): string {
  const env = Deno.env.get("PKM_USAGE_DIR");
  if (env) return env;
  return `${Deno.env.get("HOME")}/GITHUB/ObsidianPKM/skripte/usage`;
}

interface DailyRow {
  date: string;
  cost: number;
}

async function loadDailyCosts(dir: string): Promise<DailyRow[]> {
  const rows: DailyRow[] = [];
  try {
    for await (const e of Deno.readDir(dir)) {
      if (!e.isFile || !e.name.endsWith(".json")) continue;
      if (e.name.endsWith("_rl.json") || e.name.endsWith(".migrated")) continue;
      const text = await readText(join(dir, e.name));
      if (!text) continue;
      let data: Record<string, unknown>;
      try {
        data = JSON.parse(text) as Record<string, unknown>;
      } catch {
        continue;
      }
      if (typeof data !== "object" || "ccusage_daily" in data) continue;
      const d = String(data.date ?? "");
      if (!d) continue;
      const ccusage = (data.ccusage as Record<string, unknown>) ?? {};
      const cost = parseFloat(String(ccusage.total_cost ?? "0")) || 0;
      rows.push({ date: d, cost });
    }
  } catch { /* dir not readable */ }
  return rows;
}

interface RlSnapshot {
  ts?: string;
  five_hour_pct?: number;
  five_hour_resets_at?: number;
  seven_day_pct?: number;
  seven_day_resets_at?: number;
}

async function latestRlSnapshot(dir: string): Promise<RlSnapshot> {
  let bestTs = "";
  let best: RlSnapshot = {};
  try {
    for await (const e of Deno.readDir(dir)) {
      if (!e.isFile || !e.name.endsWith("_rl.json")) continue;
      const text = await readText(join(dir, e.name));
      if (!text) continue;
      let data: RlSnapshot;
      try {
        data = JSON.parse(text) as RlSnapshot;
      } catch {
        continue;
      }
      const ts = String(data.ts ?? "");
      if (ts > bestTs) {
        bestTs = ts;
        best = data;
      }
    }
  } catch { /* ok */ }
  return best;
}

interface RlInfo {
  used_pct: number;
  elapsed_pct: number;
  projected_pct: number;
  resets_in: string;
}

function rateLimitFromSnapshot(
  pct: number | undefined,
  resetsAt: number | undefined,
  windowSecs: number,
  nowTs: number,
): RlInfo | null {
  if (pct === undefined || resetsAt === undefined) return null;
  const remaining = resetsAt - nowTs;
  if (remaining <= 0 || remaining > windowSecs) return null;
  const elapsed = windowSecs - remaining;
  if (elapsed <= 0) return null;
  const frac = elapsed / windowSecs;
  const projected = frac > 0 ? pct / frac : 0;
  const d = Math.floor(remaining / 86400);
  const rem1 = remaining % 86400;
  const h = Math.floor(rem1 / 3600);
  const m = Math.floor((rem1 % 3600) / 60);
  const countdown = d ? `${d}d${h}h` : (h ? `${h}h${m}m` : `${m}m`);
  return {
    used_pct: Math.round(pct * 10) / 10,
    elapsed_pct: Math.round(frac * 1000) / 10,
    projected_pct: Math.round(projected * 10) / 10,
    resets_in: countdown,
  };
}

export async function collectCost(dir?: string): Promise<Record<string, unknown>> {
  try {
    const udir = dir ?? usageDir();
    try {
      await Deno.stat(udir);
    } catch {
      return { available: false, reason: `usage/ nicht gefunden: ${udir}` };
    }
    const rows = await loadDailyCosts(udir);
    if (!rows.length) {
      return { available: false, reason: `keine ccusage-JSON in ${udir}` };
    }

    const now = new Date();
    const todayStr = now.toISOString().slice(0, 10);
    const yestStr = new Date(now.getTime() - 86400000).toISOString().slice(0, 10);
    // Week starts Friday (Mon=0..Sun=6, Fri=4 in JS Sun=0, so Fri=5)
    const dow = now.getDay(); // 0=Sun..6=Sat
    const daysFromFri = (dow - 5 + 7) % 7;
    const weekStart = new Date(now.getTime() - daysFromFri * 86400000);
    const weekStr = weekStart.toISOString().slice(0, 10);
    const monthStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-01`;

    let todayC = 0, yestC = 0, weekC = 0, monthC = 0, totalC = 0;
    for (const r of rows) {
      totalC += r.cost;
      if (r.date === todayStr) todayC += r.cost;
      if (r.date === yestStr) yestC += r.cost;
      if (r.date >= weekStr) weekC += r.cost;
      if (r.date >= monthStr) monthC += r.cost;
    }

    const result: Record<string, unknown> = {
      available: true,
      today: todayC,
      yesterday: yestC,
      week: weekC,
      month: monthC,
      total: totalC,
      five_hour: null,
      seven_day: null,
    };

    const rl = await latestRlSnapshot(udir);
    if (rl.ts) {
      const nowTs = Math.floor(now.getTime() / 1000);
      result.five_hour = rateLimitFromSnapshot(
        rl.five_hour_pct,
        rl.five_hour_resets_at,
        5 * 3600,
        nowTs,
      );
      result.seven_day = rateLimitFromSnapshot(
        rl.seven_day_pct,
        rl.seven_day_resets_at,
        7 * 24 * 3600,
        nowTs,
      );
    }

    return result;
  } catch (exc) {
    return { available: false, reason: `collect_cost failed: ${exc}` };
  }
}
